src/service/layers-impl.js

```diff
@@ -0,0 +1,876 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {layoutRectLtwh} from '../layout-rect';
+import {dev} from '../log';
+import {filterSplice} from '../utils/array';
+import {Services} from '../services';
+import {registerServiceBuilderForDoc} from '../service';
+import {computedStyle} from '../style';
+import {listen} from '../event-helper';
+import {getMode} from '../mode';
+
+const LAYOUT_PROP = '__AMP_LAYOUT';
+
+/**
+ * The Size of an element.
+ *
+ * @struct
+ * @typedef {{
+ *   height: number,
+ *   width: number,
+ * }}
+ */
+export let SizeDef;
+
+/**
+ * The offset Position of an element.
+ *
+ * @struct
+ * @typedef {{
+ *   left: number,
+ *   top: number,
+ * }}
+ */
+export let PositionDef;
+
+/**
+ * Creates a Size.
+ *
+ * @param {number} width
+ * @param {number} height
+ * @return {!SizeDef}
+ */
+function sizeWh(width, height) {
+  return {
+    height,
+    width,
+  };
+}
+
+/**
+ * Creates a Position.
+ *
+ * @param {number} left
+ * @param {number} top
+ * @return {!PositionDef}
+ */
+function positionLt(left, top) {
+  return {
+    left,
+    top,
+  };
+}
+
+/**
+ * The core class behind the Layers system, this controls layer creation and
+ * management.
+ */
+export class LayoutLayers {
```

Comment:
Echoing Dima's comment, I think we can have a clearer naming scheme/API. 

- There's two different functions named `getOffsetPosition` which is cached as `position_` (no "offset"), which is distinct from `getOffsetFromParent`. 
- There are three different functions named `declareLayer`. 
- `LayoutLayers` is a service that manages `LayoutElements`, each of which may or may not be a "layer".

This means it takes a lot longer for new readers to get familiarized. Can you add the public functions and their doc comments for each class to your design doc? That way we can get a birds-eye view of the API.